[Nay Lin] iP

NayLin-H99 opened the pull request:
### DukePro

> "All you need is the plan, the road map, and the courage to press on to your desination." - Earl Nightingale ([Source](https://www.goodreads.com/quotes/606982-all-you-need-is-the-plan-the-road-map-and#:~:text=Quotes%20%3E%20Quotable%20Quote-,%E2%80%9CAll%20you%20need%20is%20the%20plan%2C%20the%20road%20map%2C,press%20on%20to%20your%20destination.%E2%80%9D))

DukePro frees your mind of having to remember things you need to do. It's, 

- text-based
- easy to learn
- **~~FAST~~** ***SUPER* FAST** to use

All you need to do is,

1. download it from [here](https://github.com/NayLin-H99/ip/releases/tag/A-Jar).
2. double-click it
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing Tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too! Here's the ``HelloWorld`` code:
```java
// Your First Program

class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, World!"); 
    }
}
```

Commit:
Add Gradle support

NayLin-H99 committed:
Added Greeting, echoes user input, and exits on 'bye' user input.

NayLin-H99 committed:
Abstracted out String formatting, implemented user input storage and listing when requested

NayLin-H99 committed:
Bug-fixed for Level-2

NayLin-H99 committed:
Added Task and TaskStorage to manage tasks, added ability to mark tasks as done.

NayLin-H99 committed:
Added support for more types of Tasks (ToDo, Event, Deadline)

NayLin-H99 committed:
Bug-fixed and adjusted some String formatting for Level-4

NayLin-H99 committed:
Text UI Testing for Level 1 to 4

NayLin-H99 committed:
Added DukeException, and error handling for invalid input/parameters from the user. Also removed ability to create generic Tasks, and turned Task abstract.

NayLin-H99 committed:
Added error checking for no input for "done". Test cases updated.

NayLin-H99 committed:
String formatting updated to use tab instead of whitespace.

NayLin-H99 committed:
Updated to support delete. Storage updated to use ArrayList. Minor
adjustments to string outputs and errors.

NayLin-H99 committed:
Minor adjustments.

NayLin-H99 committed:
Minor adjustment

NayLin-H99 committed:
Adding error checking for list (for empty list)

NayLin-H99 committed:
Revert "Adding error checking for list (for empty list)"

This reverts commit 654438d9314f2796a1a87d85bf1c61d6dee1c374.

NayLin-H99 committed:
Revert "Minor adjustment"

This reverts commit 449d874aeea10c6d7b59a25f6f0afdf5b5cfbe4b.

NayLin-H99 committed:
Revert "Minor adjustments."

This reverts commit c9165c3f85bc27cbe4ada66b5db35d6d8c32bd69.

NayLin-H99 committed:
Tidied up code and implemented stricter command restrictions. Added a few minor bugfixes and String output adjustments.

NayLin-H99 committed:
Added more tests to test for stricter command filters.

NayLin-H99 committed:
Adjusted strings from TaskStorage error checking for clarity.

NayLin-H99 committed:
Abstracted majority of the workflow from Duke to DukeInputProcessor and DukeOperator, with enum Operation. Used regex to clean up code and reduce throwing of exceptions.

NayLin-H99 committed:
Updated test files to the latest version of Duke

NayLin-H99 committed:
Implement new functionality to save tasks
1. Add an abstract method databaseString() to print strings to store in the database, and implement for all subclasses.
2. Add a new operation under Operation enum 'start' to start up Duke by reading data and printing the start string.

NayLin-H99 committed:
Add functionality to read date (with or without time) for tasks
1)  Adjust Event, Deadline and Task to support datetime recognition
2) Update DukeOperator operations
3) Formatted some output strings in storage

NayLin-H99 committed:
Merge branch 'branch-Level-7'

NayLin-H99 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/DukeOperator.java
#	src/main/java/Event.java

NayLin-H99 committed:
Change to utilise more OOP to suit the specifications

1) Delete DukeOperator, abstract individual commands to classes AddCommand, DeleteCommand, ExitCommand, ListCommand, DoneCommand.
2) Add enum AddCommandType to support AddCommand for different Task types.
3) Delete enum Operation.
4) Rename DukeInputProcessor to Parser. Parser now outputs a Command instead of an enum Operation.
5) Abstract Storage (loading and saving from savefile) from DukeOperator.
6) Rename TaskStorage to TaskList. TaskList operations do not return a confirmation string anymore.

NayLin-H99 committed:
Organise classes into packages
1) Create packages duke.
2) Create subpackages duke.task and duke.command.

NayLin-H99 committed:
Minor bugfix for filepath issues with saving
1) "./data/duke.txt" filepath relies on data folder preexisting. Bugfix to remove reliance.

NayLin-H99 committed:
Add JUnit test files to test Parser, Storage and Task types.

NayLin-H99 committed:
Add JavaDoc comments for all Java files.

NayLin-H99 committed:
Reformat and clean up code to match coding standard.

NayLin-H99 committed:
Add new find functionality to search tasks.

NayLin-H99 committed:
Merge branch 'A-JavaDoc'

NayLin-H99 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/DukeException.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/UI.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DoneCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java

NayLin-H99 committed:
Merge branch 'branch-Level-9'
Minor adjustments to formatting
# Conflicts:
#	src/main/java/duke/Parser.java

NayLin-H99 committed:
Check and clean code using checkstyle from AddressBook Level 3 project.

NayLin-H99 committed:
Merge branch 'branch-A-CheckStyle'

NayLin-H99 committed:
Refactor UI to Ui and rename a constant.

NayLin-H99 committed:
Reformat test files and add checkstyle file.

NayLin-H99 committed:
Merge branch 'branch-A-Gradle'

LZher385 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,151 @@
+package duke;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import duke.command.AddCommand;
+import duke.command.AddCommandType;
+import duke.command.Command;
+import duke.command.DeleteCommand;
+import duke.command.DoneCommand;
+import duke.command.ExitCommand;
+import duke.command.FindCommand;
+import duke.command.ListCommand;
+import duke.task.TaskList;
```
I like how neat the import statements are!

NayLin-H99 replied:
Thank you! 😃

LZher385 reviewed (commented):
Good job! I really struggled to find any violations 👍 . Overall, I really like how neat your code is and also the extensive  amount of tests!

jennibearduit commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,95 @@
+public class Ui {
+
+    /** line separator */
+    private static final String LINE_BREAK = "\t____________________________________________________________\n";
```
I like that you made the line break a constant field, it makes the parts where you use line break look so much more readable

NayLin-H99 replied:
Thank you! 😃

jennibearduit commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,151 @@
+     *
+     * @return boolean.
+     */
+    private boolean checkBye() {
```
According to the coding standard, boolean variables/methods should be named to sound like booleans. Perhaps `isBye()` or `isByeKeyword()` will be a better method name?

NayLin-H99 replied:
Adjusted all the methods to fit the coding standards. Thanks!

jennibearduit commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,151 @@
+     *
+     * @return boolean.
+     */
+    private boolean checkList() {
```
Similar to the `checkBye()` comment, perhaps can consider renaming the method to `isListKeyword()`? Same for the subsequent ones.

NayLin-H99 replied:
Thanks for pointing this out! I've adjusted accordingly. 👍

jennibearduit commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @return string with delimiters for saving and loading.
+     */
+    public abstract String databaseString();
```
According to the coding standard, names representing methods must be verbs. Perhaps can consider naming to `generateDatabaseString()` or perhaps `toDatabaseString()` instead, similar to Java's `toString()` method?

NayLin-H99 replied:
Done! Thanks for pointing this out as well 👍

jennibearduit commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,88 @@
+        String[] inputs = this.userInput.split(" ", 2);
+        try {
+            switch (this.addType) {
+            case todo: {
+                newTask = new Todo(inputs[1]);
+                break;
+            }
+            case event: {
+                String[] args = inputs[1].split(" /at ", 2);
+                String[] datetimeArgs = args[1].split(" ", 2);
+                if (datetimeArgs.length == 1) {
+                    newTask = new Event(args[0], args[1]);
+                } else {
+                    newTask = new Event(args[0], datetimeArgs[0], datetimeArgs[1]);
+                }
+                break;
+            }
+            case deadline: {
+                String[] args = inputs[1].split(" /by ", 2);
+                String[] datetimeArgs = args[1].split(" ", 2);
+                if (datetimeArgs.length == 1) {
+                    newTask = new Deadline(args[0], args[1]);
+                } else {
+                    newTask = new Deadline(args[0], datetimeArgs[0], datetimeArgs[1]);
+                }
+                break;
+            }
```
According to the coding standard, the format of a case block does not have curly braces, and only requires the colon. Your other switch block in `Storage.java` line 58 is correct, I would suggest following that as well.
<img width="1290" alt="Screenshot 2021-09-02 at 10 09 52 PM" src="https://user-images.githubusercontent.com/77873928/131858937-afd97184-e237-463c-b992-9b85c7775867.png">

NayLin-H99 replied:
I initially included blocks because switch case does not allow same variables to be declared in different case, and putting a block fixes that. I'll try abstracting the variables out then 😅

jennibearduit commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,95 @@
+     * @param ind index of the task to get from storage.
+     * @return the task.
+     */
+    public Task get(int ind) {
```
This is minor and subjective, but I think replacing `ind` with `index` will make the code more readable, and does not significantly change the length of the variable name.

NayLin-H99 replied:
Adjusted it to a more conventional idx to make it more readable. Thanks pointing this out 👍

jennibearduit reviewed (commented):
LGTM, code is very easy to read and understand overall!

NayLin-H99 committed:
Add JavaFX GUI support.

NayLin-H99 committed:
Merge branch 'branch-Level-10'

NayLin-H99 committed:
Fix code based on comments.

NayLin-H99 committed:
Add Assertions

NayLin-H99 committed:
Fix bug for AddCommand and TaskList

NayLin-H99 committed:
Refactor code and add Javadoc in Parser, Storage, Tasklist and commands

Refactor user input processing from commands to Parser for better abstraction.
Refactor task creation to additional methods in AddCommand for readability.
Refactor task recreation to additional methods in Storage for readability.
Add and adjust Javadoc for most classes.

NayLin-H99 committed:
Use Streams for FindCommand and TaskList

NayLin-H99 committed:
Merge pull request #2 from NayLin-H99/branch-A-CodeQuality

Branch-A-CodeQuality

NayLin-H99 committed:
Merge branch 'master' into branch-A-Streams

NayLin-H99 committed:
Merge pull request #3 from NayLin-H99/branch-A-Streams

Branch-A-Streams

NayLin-H99 committed:
Merge branch 'master' into branch-A-Assertions

NayLin-H99 committed:
Merge pull request #4 from NayLin-H99/branch-A-Assertions

Branch-A-Assertions

NayLin-H99 committed:
Add reschedule task functionality and improve regex

Add TimeTask abstract class to superclass tasks that need time or date constraints.
Adjust Event and Deadline to extend TimeTask.
Add RescheduleCommand and command recognition under Parser.
Adjust Regex under Parser for date or time constraints.
Add additional tests for new commands.

NayLin-H99 committed:
Fix Assert Statements

NayLin-H99 committed:
Merge branch 'branch-B-Snooze'

NayLin-H99 committed:
Add Update feature to allow user to update task descriptions

Add UpdateCommand to be able to change tasks descriptions.
Adjust Parser to recognize new command.
Add test cases to ParserTest.

NayLin-H99 committed:
Merge branch 'branch-C-Update'

NayLin-H99 committed:
Add default values for time in TimedTask

NayLin-H99 committed:
Improve GUI

Change color to a more eye-pleasing theme.
Clip profile picture to circle.

NayLin-H99 committed:
Update README.md

NayLin-H99 committed:
Add representative screenshot

NayLin-H99 committed:
Minor adjustment to Gradle build
